### ISSUE TITLE
(discussion) Samples

### ISSUE THREAD

**Issue by noxarivis** [mbitsnbites/soundbox#14]
theoratical it should be easy to use a "sample generator" the oscillator is recreating the soundwave of the sample instead of a sine or square.
technically it should be possible too because the oscillator plus the filters create a wave in the end.

my logic: 
(osc) ==> (wav)
(wav) ==> (osc)

adding a fiveth option next to sin, tri, saw, sqr which "generates" the sample content of the uploaded wav.

this also shouldnt break loading old songs and including it in a js demo (via base64) should also be possible.

the Efflux tracker is able to draw samples btw
https://github.com/igorski/efflux-tracker

**Comment by esaruoho** [comment]
I really wouldn't mind being able to use wavetables (wav as osc) in soundbox.

**Comment by noxarivis** [comment]
another plus by this, 
being able to draw samples...

**Comment by mbitsnbites** [comment]
Sample support is something that is incompatible with the main goal of SoundBox: producing tunes that fit in a 4 KB demo (i.e. the compressed song data + player routine must be less than 1.5 KB or so). This same goal also enables other things, such as sharing the complete song via an URL, etc.

Technically, adding support for samples is quite easy, but it would be a different product. Thus, if anyone wants to do it, my recommendation would be to make a fork of SoundBox.

**Comment by noxarivis** [comment]
it wouldnt loose its scope, just extending it.

adding samples doesnt mean we have to get rid of the synth to create 4K demos.

the user should decide if he wants to use samples or not, it would be still possible to use only the synths

**Comment by noxarivis** [comment]
to be honest, if you really think that SoundBox needs to be forked to add such things, its probably easier to port MilkyTracker to the web...

**Comment by mbitsnbites** [comment]
So, the thing is that sample support is not part of *my* focus, and while it is true that it could be an extension, it would require some extra architecturing around it. For instance:
* The player routine would have to be dynamically generated to drop the sample playback logic if samples are not used.
* Samples would have to be included in the exported JavaScript, e.g. as base64 encoded strings.
* How about really large samples? OGG files? Should you be allowed to source them via URL:s rather than raw data? How about data exchange etc?
* Some functionality may not support samples (e.g. URL-encoded songs).

Also, a hypothetical online song database (an idea I'm toying with) would be much less feasible if songs are megabytes in size each rather than 0.5 KB on average.

In conclusion: Just adding sample support to the player routine is near-trivial, and adding it to the UI would probably not be much work. But to get the logistics right requires some work. And that work is not something that I am personally ready to prioritize. However, if someone wanted to shift focus and move quickly, a fork would be a feasible solution IMO.

**Comment by esaruoho** [comment]
> Sample support is something that is incompatible with the main goal of SoundBox: producing tunes that fit in a 4 KB demo

@mbitsnbites 

hi.. i did see what the original poster suggested which was this (in my understanding)

you create a pattern for one preset.
you render that pattern into wav.
you load that wav into an oscillator, and then play another pattern with it.  <-kind of like nested diskwrites, done as-is, to fit into the 4Kb limitation. not like "oh i saved a song on soundbox that is 300mb in size, out of scope of 4kb intro usage" :)

**Comment by mbitsnbites** [comment]
Ok, that's much more feasible, and I can see how it could be used for creating some interesting sounds (would be even more interesting if you could combine several channels with different patterns into one wave).

However, it would still complicate the player routine (even if not by much), so you would probably want to have an easy way to drop that support in a final product (e.g. just cutting out a certain piece of code from the player routine).

Might also be tricky to get a good UI around it (which channels and patterns should be "sample sources", and how would interactive playback work?).

**Comment by mbitsnbites** [comment]
On a related note, I tried to add recursive wave generation to the [Spectre synth](http://spectre.bitsnbites.eu/editor/), which is already completely wave-based btw, but the player routine got too large.
